### PR TITLE
New Apache Range Header DoS Scanner

### PR DESCRIPTION
**Pull request by kingthorin**
Add scanner for Apache Range Header DoS (CVE-2011-3192). Includes help
updates, and ZapAddOn.xml updates.

### CODE REVIEW

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+			try {
+				sendAndReceive(newRequest, false);
+			} catch (IOException e) {
+				LOG.warn("An error occurred while checking [" + newRequest.getRequestHeader().getMethod() + "] ["
+						+ newRequest.getRequestHeader().getURI() + "] for Apache Range Header DoS (CVE-2011-3192)."
+						+ "Caught " + e.getClass().getName() + " " + e.getMessage());
+				return;
+			}
+
+			if (newRequest.getResponseHeader().getStatusCode() == HttpStatusCode.PARTIAL_CONTENT) {// 206
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: Is the comment really needed?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+		HttpMessage chkRequest = getNewMsg();
+		chkRequest = setRequestHeaders(chkRequest, "0-6");
+
+		try {
+			sendAndReceive(chkRequest, false);
+		} catch (IOException e) {
+			LOG.warn("An error occurred while validating [" + chkRequest.getRequestHeader().getMethod() + "] ["
+					+ chkRequest.getRequestHeader().getURI()
+					+ "] for Apache Range Header DoS (CVE-2011-3192) applicability." + "Caught "
+					+ e.getClass().getName() + " " + e.getMessage());
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: Could return false immediately, no need to check the status code.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+		}
+
+		if (chkRequest.getResponseHeader().getStatusCode() == HttpStatusCode.PARTIAL_CONTENT) {// 206
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: Could return the value of the comparison instead.
The comment could be removed?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+			return true;
+		} else {
+			return false;
+		}
+	}
+
+	private HttpMessage setRequestHeaders(HttpMessage aMessage, String rangeValue) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: Should the method return? The message passed as parameter is already being changed.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+		aMessage.getRequestHeader().setHeader("Range", "bytes=" + rangeValue);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: The concatenation of the string could be done once and used by both calls.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/Messages.properties [hunk 6]
```diff
@@ -3,6 +3,11 @@
 # This file defines the default (English) variants of all of the internationalised messages
 ascanalpha.desc = Active Scan Rules - alpha
 
+ascanalpha.apacherangeheaderdosscanner.name = Apache Range Header DoS (CVE-2011-3192)
+ascanalpha.apacherangeheaderdosscanner.desc = The byterange filter in earlier versions of the Apache HTTP Server allows remote attackers to cause a denial of service (memory and CPU exhaustion) via a Range request header that identifies multiple overlapping ranges. This issue was exploited in the wild in August 2011.
+ascanalpha.apacherangeheaderdosscanner.soln = Upgrade your Apache server to a currently stable version. Alternatives solutions or workarounds are outlined in the references.
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/Messages.properties]: "Alternative solutions" instead?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/help/contents/ascanalpha.html [hunk 7]
```diff
@@ -9,6 +9,10 @@
 <H1>Active Scan Rules - alpha</H1>
 The following alpha quality active scan rules are included in this add-on:
 
+<H2>Apache Range Header DoS (CVE-2011-3192)</H2>
+Tests to see if the server is vulnerable to the Apache Range Header Denial of Service issue. By requesting 11 different byte ranges, one more than should be
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/help/contents/ascanalpha.html]: "than what"?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+				bingo(getRisk(), // Risk
+						Alert.CONFIDENCE_MEDIUM, // Confidence/Reliability
+						getName(), // Name
+						getDescription(), // Description
+						newRequest.getRequestHeader().getURI().toString(), // URI
+						null, // Param
+						"", // Attack
+						"", // OtherInfo
+						getSolution(), // Solution
+						"", // Evidence
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: Could use the status line (`getPrimeHeader()`) as evidence?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/help/contents/ascanalpha.html (hunk 7 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/help/contents/ascanalpha.html]: It's on the next line: 

> ... one more than should be accepted by a patched/fixed server.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java (hunk 1 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: Nope.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java (hunk 2 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: I'm not sure what you mean here. I can't return false from the try/catch as the method doesn't return a boolean. The return inside the catch will exit the function won't it? The same as the return inside the isStop() check at L122?

Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: Yup

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/help/contents/ascanalpha.html (hunk 7 above)
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/help/contents/ascanalpha.html]: Yeah, I read that as well, the "more than should be" didn't sound good, if it's how it should be I'm ok with  that.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java (hunk 3 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: Yup

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java (hunk 4 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: I guess it doesn't need to.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java (hunk 5 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: Good point

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/Messages.properties (hunk 6 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/Messages.properties]: Yup, oops.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/help/contents/ascanalpha.html (hunk 7 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/help/contents/ascanalpha.html]: Hmm ok I'll come up with a different way to say it.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java (hunk 8 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ApacheRangeHeaderDos.java]: Sure

**Comment by thc202** [comment]
Looks good to me, thank you.

**Pull request merged**